wit/types.wit

```diff
@@ -244,7 +244,23 @@ interface types {
 
     /// Gives the `incoming-body` associated with this request. Will only
     /// return success at most once, and subsequent calls will return error.
+    ///
+    /// Deprecated in favor of `into-parts`.
     consume: func() -> result<incoming-body>;
+
+    /// Decompose the incoming request into its component parts.
+    ///
+    /// All resources returned by this function are owned.
+    into-parts: static func(
+      response: incoming-request
+    ) -> tuple<
+      method,
+      option<string>, // path-with-query
+      option<scheme>,
+      option<string>, // authority
+      headers,
+      incoming-body
+    >;
```

Comment:
Given that one can already copy everything except the `headers` and `incoming-body` fields, I wonder if we really need a method that copies all fields at once since it may often be the case that a number of the fields will go unused (which won't be easily optimized away).

Instead, given that `consume` already exists in 0.2 and isn't going anywhere, I'd suggest we make the minimal addition in 0.2.1 and just add a `take-headers`, and then consider our next major refactoring of the design as part of 0.3.

Reply:
That seems reasonable. To make the situation more generic, what we're really after are ways to take ownership of child resources in order to avoid deep-copying what those resources represent if they are needed elsewhere. Through that lens, the `variant`s and `string`s are indeed superfluous.

I think `take-headers` suffices for now since this PR is only augmenting the `incoming-` types which cannot be sent anywhere. It might be worth thinking about the behavior of such a method once incoming/outgoing are unified, e.g. what does it mean to send a request whose headers have been taken? There are a number of reasonable behaviors we might pick, such as the headers starting over from empty after being taken, or returning an error if trying a send-ater-take. A single static method that returns all of the child resources could be appealing to resolve that ambiguity.

Reply:
Thanks!